autogen/agentchat/contrib/nexusravenv2_local_function_calling.py

```diff
@@ -0,0 +1,162 @@
+                     {'type': 'function',
+                     'function': {'description': 'This is a random word generator.',
+                                  'name': 'random_word_generator',
+                                  'parameters': {'type': 'object',
+                                                 'properties': {'seed':   {'type': 'integer',
+                                                                           'description': 'seed for initialising the random generator'},
+                                                                'prefix': {'type': 'string',
+                                                                           'description': 'string to prefix random word'}},
+                                                 'required': ['seed', 'prefix']}}}],
+             'model': 'nexusraven'}]"""
+    if getattr(func, '_is_decorated', False):
+        # If the function is already decorated, return it as is
+        return func
+
+    def wrapper(*args, **kwargs):
+        result = func(*args, **kwargs)
+
+        # extract the tools element from the result
+        query_header = "\n\n".join(
+            [create_nexus_prompt_for_tool(tool) for tool in result['tools'] if tool['type'] == 'function'])
+
+        query = "\n".join([m['content'] for m in result['messages'][-1:]])
+        prompt = f"""<human>:\n{query_header}\nUser Query: Question: {query} \n suggest a function and suitable argument values for the situation conversation:  """
+        del result['messages']
+        del result['tools']
+        result['prompt'] = prompt
+
+        result['extra_query'] = {"inputs": prompt,
+                                 "parameters": {"temperature": 0.001, "do_sample": False, "max_new_tokens": 2000}}
+        result['extra_body'] = {"prompt": prompt}
+        return result
+
+    wrapper._is_decorated = True
+    return wrapper
+
+
+class NexusFunctionCallingAssistant(autogen.AssistantAgent):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.register_reply([autogen.Agent, None], NexusFunctionCallingAssistant.generate_oai_reply)
+
+    @staticmethod
+    def parse_function_details(input_string: str) -> Tuple[str, Dict[str, str], str] | None:
+        call_part, thought_part = input_string.split("<bot_end> \nThought: ")
+        function_name_match = re.search(r'Call: (\w+)', call_part)
+        function_name = function_name_match.group(1) if function_name_match else None
+        args_match = re.search(r'\((.*?)\)', call_part)
+        args_str = args_match.group(1) if args_match else ''
+        args_list = args_str.split(', ')
+        args_map = {}
+        for arg in args_list:
+            key, value = arg.split('=')
+            args_map[key.strip()] = value.strip().strip('\'')
+        return function_name, args_map, thought_part.strip()
+
+    @override
+    def _generate_oai_reply_from_client(self, llm_client, messages, cache) -> Union[str, Dict, None]:
```

Comment:
An abstraction leakage here. We might change the private method here in the future. Perhaps a complete reply function instead of overriding?

Reply:
Also type hints for the arguments.